qiskit/qobj/qasm_qobj.py

```diff
@@ -0,0 +1,482 @@
+        self.___dict___['parameter_binds'] = value
+
+    @property
+    def memory_slots(self):
+        """The number of memory slots on the device."""
+        self.__dict__.get('memory_slots')
+
+    @memory_slots.setter
+    def memory_slots(self, value):
+        self.__dict__['memory_slots'] = value
+
+    @property
+    def n_qubits(self):
+        """The number of qubits on the device."""
+        nqubits = self.__dict__.get('n_qubits')
+        if nqubits is None:
+            raise AttributeError('Attribute n_qubits is not defined')
+        return nqubits
+
+    @n_qubits.setter
+    def n_qubits(self, value):
+        self.__dict__['n_qubits'] = value
+
+    def to_dict(self):
+        """Return a dictionary format representation of the QASM Qobj config.
+
+        Returns:
+            dict: The dictionary form of the QasmQobjConfig.
+        """
+        return self.__dict__
+
+    @classmethod
+    def from_dict(cls, data):
+        """Create a new QasmQobjConfig object from a dictionary.
+
+        Args:
+            data (dict): A dictionary for the config
+
+        Returns:
+            QasmQobjConfig: The object from the input dictionary.
+        """
+        return cls(**data)
+
+    def __eq__(self, other):
+        if isinstance(other, QasmQobjConfig):
+            if self.to_dict() == other.to_dict():
+                return True
+        return False
+
+
+class QobjDictField(SimpleNamespace):
+    """A class used to represent a dictionary field in Qobj
+
+    Exists as a backwards compatibility shim around a dictionary for Qobjs
+    previously constructed using marshmallow.
+    """
+
+    def __init__(self, **kwargs):
+        """Instantiate a new Qobj dict field object.
+
+        Args:
+            kwargs: arbitrary keyword arguments that can be accessed as
+                attributes of the object.
+        """
+        self.__dict__.update(kwargs)
+
+    def to_dict(self):
+        """Return a dictionary format representation of the QASM Qobj.
+
+        Returns:
+            dict: The dictionary form of the QobjHeader.
+
+        """
+        return self.__dict__
+
+    @classmethod
+    def from_dict(cls, data):
+        """Create a new QobjHeader object from a dictionary.
+
+        Args:
+            data (dict): A dictionary representing the QobjHeader to create. It
+                will be in the same format as output by :func:`to_dict`.
+
+        Returns:
+            QobjDictFieldr: The QobjDictField from the input dictionary.
+        """
+
+        return cls(**data)
+
+    def __eq__(self, other):
+        if isinstance(other, self.__class__):
+            if self.__dict__ == other.__dict__:
+                return True
+        return False
+
+
+class QasmQobjExperimentHeader(QobjDictField):
+    """A header for a single QASM experiment in the qobj."""
+    pass
+
+
+class QasmQobjExperimentConfig(QobjDictField):
+    """Configuration for a single QASM experiment in the qobj."""
+    pass
```

Comment:
Without explicitly declaring the attributes, auto-complete won't pick up the attributes. Not sure how important it is for this class (and the others) though.

Reply:
In my opinion, these fields are used quite a bit less.